common/range-map/src/range_map.rs

```diff
@@ -0,0 +1,255 @@
+// Copyright 2021 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use core::cmp::Ordering;
+use core::fmt::Debug;
+use core::fmt::{self};
+use core::ops::Bound;
+use core::ops::Range;
+use std::collections::BTreeMap;
+
+use super::range_wrapper::RangeWrapper;
+
+#[derive(Clone)]
+pub struct RangeMap<K, V> {
+    // Wrap ranges so that they are `Ord`.
+    // See `range_wrapper.rs` for explanation.
+    pub(crate) btm: BTreeMap<RangeWrapper<K>, V>,
+}
+
+impl<K, V> Default for RangeMap<K, V>
```

Comment:
I do not quite get the design goal of this map:

- Does it allow to insert overlapping `range` with different values? If it does allow this to happen, what will a `get()` actually get?
- Is a query for a sub-range allowed? What if the range map contains two ranges `[1,3]` and `[3,5]` and a query for `[2,4]`? Will it return a positive result?